client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -203,8 +214,41 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
-        Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, config);
-        synchronizer.updateStateUnconditionally(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(config)));
+        while (true) {
+            synchronizer.fetchUpdates();
+            val state = synchronizer.getState();
+            // We only move into the block if the state transition has happened successfully.
+            if (stateTransition(state, new UpdatingConfig(true))) {
+                // Use the latest generation and reader group Id.
+                ReaderGroupConfig newConfig = config.toBuilder()
+                        .readerGroupId(state.getConfig().getReaderGroupId())
+                        .generation(state.getConfig().getGeneration()).build();
+                boolean success = Futures.getThrowingException(controller.updateReaderGroup(scope, groupName, newConfig));
+                if (success) {
+                    synchronizer.updateState((s, updates) -> {
+                        Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, newConfig);
+                        updates.add(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(newConfig), false));
+                    });
+                    return;
+                }
+            }
+        }
+    }
+
+    private boolean stateTransition(ReaderGroupState state, ReaderGroupState.ReaderGroupStateUpdate update) {
+        // This boolean will help know if the update actually succeeds or not.
+        AtomicBoolean successfullyUpdated = new AtomicBoolean(true);
+        synchronizer.updateState((s, updates) -> {
+            // If successfullyUpdated is false then that means the current state where this update should
+            // take place (i.e. state with updatingConfig as false) is not the state we are in so we do not
+            // make the update.
+            boolean updated = s.getConfig().equals(state.getConfig());
```

Comment:
State is only used for config. So make the parameter 'config' instead. This would also minorly simplify the caller.

Reply:
Using config instead.